cdap-api/src/main/java/co/cask/cdap/api/dataset/table/RowColumns.java

```diff
@@ -29,50 +31,127 @@
-
-  public byte[] getRow() {
-    return row;
-  }
-
-  public List<byte[]> getColumns() {
-    return columns;
-  }
+  private List<byte[]> columns;
 
   // key & column as byte[]
 
+  /**
+   * Defines a row with all columns included. If any of the {@code add} method is called, only columns specified through
+   * those {@code add} method calls are included.
+   * @param row the row to define.
+   */
   public RowColumns(byte[] row) {
     this.row = row;
-    this.columns = new ArrayList<>();
   }
 
+  /**
+   * Defines a row, with the specified set of columns.
+   * @param row the row to define
+   * @param columns the set of columns of the row to included
+   */
   public RowColumns(byte[] row, byte[]... columns) {
-    this(row);
-    add(columns);
-  }
-
-  @SuppressWarnings("unchecked")
-  public T add(byte[]... columns) {
-    Collections.addAll(this.columns, columns);
-    return (T) this;
+    this.row = row;
+    this.columns = Arrays.asList(columns);
   }
 
   // key and column as String
 
+  /**
+   * Defines a row with all columns included. If any of the {@code add} method is called, only columns specified through
+   * those {@code add} method calls are included.
+   * @param row the row to define. Note: It will be converted to {@code byte[]} with UTF-8 encoding.
+   */
   public RowColumns(String row) {
     this(Bytes.toBytes(row));
   }
 
+  /**
+   * Defines a row, including all columns, unless additional columns are specified.
+   * Note that the provided strings will be converted to {@code byte[]} with UTF-8 encoding.
+   *
+   * @param row the row to define
+   * @param columns the set of columns of the row to included
+   */
   public RowColumns(String row, String... columns) {
     this(row);
     add(columns);
   }
 
+  /**
+   * Specifies column(s) to be included.
+   * @param firstColumn the first column being included
+   * @param moreColumns a collection of additional columns being included
+   * @return the {@link RowColumns} object being defined
+   */
+  @SuppressWarnings("unchecked")
+  public T add(byte[] firstColumn, byte[]... moreColumns) {
+    initArray();
+    this.columns.add(firstColumn);
+    Collections.addAll(this.columns, moreColumns);
+    return (T) this;
+  }
+
+  /**
+   * Specifies column(s) to be included.
+   * @param columns a collection of columns being included
+   * @return the {@link RowColumns} object being defined
+   */
+  @SuppressWarnings("unchecked")
+  public T add(byte[][] columns) {
+    initArray();
+    Collections.addAll(this.columns, columns);
+    return (T) this;
+  }
+
+  /**
+   * Specifies column(s) to be included.
+   * Note that the provided strings will be converted to {@code byte[]} with UTF-8 encoding.
+   * @param firstColumn the first column being included
+   * @param moreColumns a collection of additional columns being included
+   * @return the {@link RowColumns} object being defined
+   */
   @SuppressWarnings("unchecked")
-  public T add(String... columns) {
```

Comment:
Hmm, good point. Even though we do say users will have to recompile and redeploy after an upgrade, perhaps we should push this change to a non-patch version (i.e. to 3.4.0).